File: src/scanner.rs
//! The SCAN driver. Iterates the whole keyspace without ever calling KEYS
//! (which would block Redis), optionally enriching each batch with MEMORY
//! USAGE / TYPE / TTL via a single pipeline round-trip per batch.

use crate::grouping::parse_key;
use crate::stats::{KeyObservation, Stats};
use anyhow::{Context, Result};
use redis::aio::ConnectionManager;

#[derive(Clone, Copy)]
pub struct ScanOptions {
    /// COUNT hint for SCAN — keys per server-side step.
    pub count: usize,
    /// Treat first N chars as namespace when a key has no ':'.
    pub namespace_len: usize,
    /// Stop after this many keys (0 = unlimited). Used for sampling.
    pub limit: u64,
    pub want_memory: bool,
    pub want_types: bool,
    pub want_ttl: bool,
    /// Retain this many largest keys (by memory). 0 disables the leaderboard.
    pub biggest: usize,
}

/// Scan the entire keyspace, returning aggregated stats.
///
/// `on_progress` is invoked after every SCAN batch with the stats accumulated
/// so far. Callers use it to drive a progress bar (CLI) or to publish live
/// snapshots into shared state (TUI). It is a no-op for the simplest callers.
pub async fn scan<F>(
    conn: &mut ConnectionManager,
    match_pattern: Option<&str>,
    opts: ScanOptions,
    mut on_progress: F,
) -> Result<Stats>
where
    F: FnMut(&Stats),
{
    let mut stats = Stats::with_biggest_cap(opts.biggest);

    let mut cursor: u64 = 0;
    loop {
        // SCAN cursor [MATCH pattern] COUNT count
        let mut cmd = redis::cmd("SCAN");
        cmd.arg(cursor).arg("COUNT").arg(opts.count);
        if let Some(pat) = match_pattern {
            cmd.arg("MATCH").arg(pat);
        }

        let (next, batch): (u64, Vec<String>) =
            cmd.query_async(conn).await.context("SCAN command failed")?;

        if !batch.is_empty() {
            let observations = enrich(conn, &batch, opts).await?;
            for (key, obs) in batch.iter().zip(observations) {
                let parsed = parse_key(key, opts.namespace_len);
                stats.record(key, &parsed.namespace, &parsed.key_type, &obs);
            }
            on_progress(&stats);
        }

        cursor = next;
        if cursor == 0 {
            break;
        }
        if opts.limit != 0 && stats.total_keys >= opts.limit {
            break;
        }
    }

    Ok(stats)
}

/// Build one observation per key. When no enrichment is requested this is a
/// cheap no-op. Otherwise all extra commands for the whole batch go out in a
/// single pipeline so we pay one network round-trip per SCAN batch, not one
/// per key.
async fn enrich(
    conn: &mut ConnectionManager,
    keys: &[String],
    opts: ScanOptions,
) -> Result<Vec<KeyObservation>> {
    let mut observations: Vec<KeyObservation> =
        keys.iter().map(|_| KeyObservation::default()).collect();

    if !opts.want_memory && !opts.want_types && !opts.want_ttl {
        return Ok(observations);
    }

    let mut pipe = redis::pipe();
    for key in keys {
        if opts.want_memory {
            pipe.cmd("MEMORY").arg("USAGE").arg(key);
        }
        if opts.want_types {
            pipe.cmd("TYPE").arg(key);
        }
        if opts.want_ttl {
            pipe.cmd("TTL").arg(key);
        }
    }

    // The pipeline returns replies in command order. Decode generically so a
    // nil MEMORY USAGE (key gone) doesn't abort the whole batch.
    let replies: Vec<redis::Value> = pipe
        .query_async(conn)
        .await
        .context("enrichment pipeline failed")?;

    let mut idx = 0;
    for obs in observations.iter_mut() {
        if opts.want_memory {
            obs.bytes = match replies.get(idx) {
                Some(redis::Value::Int(n)) if *n >= 0 => Some(*n as u64),
                _ => None,
            };
            idx += 1;
        }
        if opts.want_types {
            obs.data_type = match replies.get(idx) {
                Some(redis::Value::SimpleString(s)) => Some(s.clone()),
                Some(redis::Value::BulkString(b)) => Some(String::from_utf8_lossy(b).into_owned()),
                _ => None,
            };
            idx += 1;
        }
        if opts.want_ttl {
            obs.ttl = match replies.get(idx) {
                Some(redis::Value::Int(n)) => Some(*n),
                _ => None,
            };
            idx += 1;
        }
    }

    Ok(observations)
}

File: internal/scanner.go
package internal

import (
	"context"
	"fmt"

	"github.com/redis/go-redis/v9"
)

// ScanOptions controls the scanning behavior.
type ScanOptions struct {
	// Count hint for SCAN — keys per server-side step.
	Count int
	// Treat first N chars as namespace when a key has no ':'.
	NamespaceLen int
	// Stop after this many keys (0 = unlimited).
	Limit int64
	// Collect memory usage per key.
	WantMemory bool
	// Collect Redis data type per key.
	WantTypes bool
	// Collect TTL/expiry stats per key.
	WantTTL bool
	// Retain this many largest keys (0 disables leaderboard).
	Biggest int
}

// OnProgress is called after every SCAN batch with the stats accumulated so far.
type OnProgress func(stats *Stats)

// Scan iterates the keyspace using SCAN (never KEYS), optionally enriching each
// batch with MEMORY USAGE / TYPE / TTL via pipelining.
func Scan(ctx context.Context, client *redis.Client, matchPattern string, opts ScanOptions, onProgress OnProgress) (*Stats, error) {
	stats := NewStats(opts.Biggest)

	var cursor uint64 = 0
	for {
		keys, nextCursor, err := client.Scan(ctx, cursor, matchPattern, int64(opts.Count)).Result()
		if err != nil {
			return nil, fmt.Errorf("SCAN command failed: %w", err)
		}

		if len(keys) > 0 {
			observations, err := enrich(ctx, client, keys, opts)
			if err != nil {
				return nil, err
			}
			for i, key := range keys {
				parsed := ParseKey(key, opts.NamespaceLen)
				stats.Record(key, parsed.Namespace, parsed.KeyType, &observations[i])
			}
			if onProgress != nil {
				onProgress(stats)
			}
		}

		cursor = nextCursor
		if cursor == 0 {
			break
		}
		if opts.Limit != 0 && stats.TotalKeys >= opts.Limit {
			break
		}
	}

	return stats, nil
}

// enrich builds one observation per key. When no enrichment is requested this
// is a cheap no-op. Otherwise all extra commands for the whole batch go out via
// pipelining.
func enrich(ctx context.Context, client *redis.Client, keys []string, opts ScanOptions) ([]KeyObservation, error) {
	observations := make([]KeyObservation, len(keys))

	if !opts.WantMemory && !opts.WantTypes && !opts.WantTTL {
		return observations, nil
	}

	pipe := client.Pipeline()
	memoryCmds := make([]*redis.IntCmd, 0, len(keys))
	typeCmds := make([]*redis.StatusCmd, 0, len(keys))
	ttlCmds := make([]*redis.DurationCmd, 0, len(keys))

	for _, key := range keys {
		if opts.WantMemory {
			memoryCmds = append(memoryCmds, pipe.MemoryUsage(ctx, key))
		}
		if opts.WantTypes {
			typeCmds = append(typeCmds, pipe.Type(ctx, key))
		}
		if opts.WantTTL {
			ttlCmds = append(ttlCmds, pipe.TTL(ctx, key))
		}
	}

	_, _ = pipe.Exec(ctx)

	for i := range observations {
		obs := KeyObservation{}

		if opts.WantMemory {
			idx := 0
			if opts.WantMemory {
				if i < len(memoryCmds) {
					val, err := memoryCmds[i].Result()
					if err == nil {
						obs.Bytes = &val
					}
				}
				idx++
			}
			_ = idx
		}
		if opts.WantTypes {
			if i < len(typeCmds) {
				val, err := typeCmds[i].Result()
				if err == nil {
					obs.DataType = &val
				}
			}
		}
		if opts.WantTTL {
			if i < len(ttlCmds) {
				val, err := ttlCmds[i].Result()
				if err == nil {
					ttl := int64(val.Seconds())
					obs.TTL = &ttl
				}
			}
		}
		observations[i] = obs
	}

	// Re-process memory observables more carefully
	mi := 0
	ti := 0
	tti := 0
	for i := range observations {
		if opts.WantMemory {
			if mi < len(memoryCmds) {
				val, err := memoryCmds[mi].Result()
				if err == nil {
					observations[i].Bytes = &val
				}
				mi++
			}
		}
		if opts.WantTypes {
			if ti < len(typeCmds) {
				val, err := typeCmds[ti].Result()
				if err == nil {
					observations[i].DataType = &val
				}
				ti++
			}
		}
		if opts.WantTTL {
			if tti < len(ttlCmds) {
				val, err := ttlCmds[tti].Result()
				if err == nil {
					ttl := int64(val.Seconds())
					observations[i].TTL = &ttl
				}
				tti++
			}
		}
	}

	return observations, nil
}
